usecases/auth/authentication/apikey/dynamic_users.go

```diff
@@ -0,0 +1,161 @@
+	IdToIdentifier       map[string]string
+	users                map[string]*User
+}
+
+func NewDynamicApiKey() *DynamicApiKey {
+	return &DynamicApiKey{
+		weakKeyStorageById:   make(map[string][sha256.Size]byte),
+		secureKeyStorageById: make(map[string]string),
+		identifierToId:       make(map[string]string),
+		IdToIdentifier:       make(map[string]string),
+		users:                make(map[string]*User),
+	}
+}
+
+func (c *DynamicApiKey) CreateUser(userId, secureHash, userIdentifier string) error {
+	c.Lock()
+	defer c.Unlock()
+	_, secureKeyExists := c.secureKeyStorageById[userId]
+	_, identifierExists := c.identifierToId[userId]
+	_, usersExists := c.users[userId]
+
+	// Todo-Dynuser: Does this make sense? Error would end up with RAFT layer
+	if secureKeyExists || identifierExists || usersExists {
+		return fmt.Errorf("user %s already exists", userId)
+	}
+	c.secureKeyStorageById[userId] = secureHash
+	c.identifierToId[userIdentifier] = userId
+	c.IdToIdentifier[userId] = userIdentifier
+	c.users[userId] = &User{Id: userId, Active: true, InternalIdentifier: userIdentifier}
+	return nil
+}
+
+func (c *DynamicApiKey) RotateKey(userId, secureHash string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	c.secureKeyStorageById[userId] = secureHash
+	delete(c.weakKeyStorageById, userId)
+	return nil
+}
+
+func (c *DynamicApiKey) DeleteUser(userId string) error {
+	c.Lock()
+	defer c.Unlock()
+
+	delete(c.secureKeyStorageById, userId)
+	delete(c.identifierToId, userId)
+	delete(c.users, userId)
+	delete(c.weakKeyStorageById, userId)
+	return nil
+}
+
+func (c *DynamicApiKey) GetUsers(userIds ...string) (map[string]*User, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	users := make(map[string]*User, len(userIds))
+	for _, id := range userIds {
+		user, ok := c.users[id]
+		if ok {
+			users[id] = user
+		}
+	}
+	return users, nil
+}
+
+func (c *DynamicApiKey) CheckUserIdentifierExists(userIdentifier string) (bool, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	_, ok := c.users[userIdentifier]
+	return ok, nil
+}
+
+func (c *DynamicApiKey) ValidateAndExtract(key, userIdentifier string) (*models.Principal, error) {
+	c.RLock()
+	defer c.RUnlock()
+
+	userId, ok := c.identifierToId[userIdentifier]
+	if !ok {
+		return nil, fmt.Errorf("invalid token")
+	}
+
+	secureHash, ok := c.secureKeyStorageById[userId]
+	if !ok {
+		return nil, fmt.Errorf("invalid token")
+	}
+	weakHash, ok := c.weakKeyStorageById[userId]
+	if ok {
+		// use the secureHash as salt for the computation of the weaker in-memory
+		return c.validateWeakHash([]byte(key+secureHash), weakHash, userId)
+	} else {
+		return c.validateStrongHash(key, secureHash, userId)
+	}
+}
+
+func (c *DynamicApiKey) validateWeakHash(key []byte, weakHash [32]byte, userId string) (*models.Principal, error) {
+	keyHash := sha256.Sum256(key)
+	if subtle.ConstantTimeCompare(keyHash[:], weakHash[:]) == 1 {
+		return &models.Principal{Username: userId}, nil
+	}
+
+	return nil, fmt.Errorf("invalid token")
+}
```

Comment:
Compared to `validateStrongHash`, the bool logic here is reversed, i.e. returning a model inside the successful if statement and `nil` at the end. Can it be swapped for consistency?